Notifying users of available add-on updates

seanbudd opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review". 
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Closes #15035

### Summary of the issue:
Users would like a push notification on NVDA start-up letting them know if add-ons have available updates

### Description of user facing changes
When starting NVDA, a dialog will appear notifying users of updatable add-ons if there are available updates.
This will only notify users if the available update is in the same channel as the installed add-on.
There is an "update all" button on the dialog and a list of add-ons with updates available.
There is also an "Open Add-on Store" button to open the add-on store to the updatable add-ons tab.

A setting to disable this. This is a combobox so that in future users can decide between:
- notify on updates (default)
- automatic update (See #3208)
- disabled

### Description of development approach
Created a scheduling module to schedule tasks on NVDA start up, using the schedule pip module.
This is so we can have conflict free scheduling. e.g. so NVDA's  update notification won't clash with the add-on update notification.
Uses their suggested code to run a background thread for task scheduling: https://schedule.readthedocs.io/en/stable/background-execution.html

Changed much of the Add-on Store downloading code to be classmethods. This allows us to use it from the message dialog without creating a store instance.

### Testing strategy:
**Manual testing**
See the manual test plan for the add-on store.

**Unit testing**
Unit testing to smoke test handling the daily job at startup

### Known issues with pull request:
None

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced automatic update notifications for add-ons.
  - Added a daily job to check for updatable add-ons.

- **Enhancements**
  - Updated Add-on Store to support updating multiple add-ons at once.
  - Improved Add-on Store dialog to open to a specific tab based on installed add-ons.

- **Documentation**
  - Updated manual with instructions for simulating add-on updates and testing automatic update notifications.

- **Tests**
  - Added unit tests for scheduling daily jobs and handling job clashes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seanbudd committed:
Create message dialog notifying users of an add-on update

seanbudd committed:
Only notify updates from the same channel

josephsl commented:
CC @XLTechie and other Add-on Updater users – testing and reviews are appreciated. Thanks.

 

From: Sean Budd ***@***.***> 
Sent: Thursday, May 30, 2024 11:02 PM
To: nvaccess/nvda ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [nvaccess/nvda] Notifying users of available add-on updates (PR #16636)

 


Link to issue number:


Closes #15035 <https://github.com/nvaccess/nvda/issues/15035> 


Summary of the issue:


Users would like a push notification on NVDA start-up letting them know if add-ons have available updates


Description of user facing changes


When starting NVDA, a dialog will appear notifying users of updatable add-ons if there are available updates.
This will only notify users if the available update is in the same channel as the installed add-on.

TODO: A setting to disable this. This should be a combobox so that in future users can decide between:

*	notify on updates (default)
*	automatic update (See #3208 <https://github.com/nvaccess/nvda/issues/3208> )
*	disabled


Description of development approach


Created a scheduling module to schedule tasks on NVDA start up.
This is so we can have conflict free scheduling. e.g. so NVDA's update notification won't clash with the add-on update notification.

Ongoing - this is a draft


Testing strategy:


Ongoing - this is a draft.
Steps should be added to the manual test plan for the add-on store.


Known issues with pull request:


Code Review Checklist:


*	[ ] Documentation: 

*	Change log entry
*	User Documentation
*	Developer / Technical Documentation
*	Context sensitive help for GUI changes

*	[ ] Testing: 

*	Unit tests
*	System (end to end) tests
*	Manual testing

*	[ ] UX of all users considered: 

*	Speech
*	Braille
*	Low Vision
*	Different web browsers
*	Localization in other languages / culture than English

*	[ ] API is compatible with existing add-ons.
*	[ ] Security precautions taken.

  _____  


You can view, comment on, or merge this pull request online at:


  https://github.com/nvaccess/nvda/pull/16636


Commit Summary


*	4b97d4a <https://github.com/nvaccess/nvda/pull/16636/commits/4b97d4a9c5b849db74671934ad035655554bdf3b>  Create message dialog notifying users of an add-on update
*	c236af8 <https://github.com/nvaccess/nvda/pull/16636/commits/c236af8ffa8268578a7ea014668ea04f4334c89f>  Only notify updates from the same channel


File Changes 


(8 <https://github.com/nvaccess/nvda/pull/16636/files>  files)

*	M requirements.txt <https://github.com/nvaccess/nvda/pull/16636/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552>  (3) 
*	M source/addonStore/dataManager.py <https://github.com/nvaccess/nvda/pull/16636/files#diff-9fd9bbb610bac3157554c0876fba6ea3475383ea9de36c2101ba1127cf5189e5>  (14) 
*	M source/core.py <https://github.com/nvaccess/nvda/pull/16636/files#diff-dbe862167add4fbe8e502757f792615f714c7603621c03d0c60c08a6a613ad68>  (5) 
*	M source/gui/__init__.py <https://github.com/nvaccess/nvda/pull/16636/files#diff-eba9c2318682ea640a71acae6a4cdb2acdb3ff4a493b7fce7547cd51e52cf234>  (15) 
*	M source/gui/addonStoreGui/__init__.py <https://github.com/nvaccess/nvda/pull/16636/files#diff-6276080b60b6eafe6906f927b768328b1e718119ae201cd40a24c153eb29cf42>  (10) 
*	M source/gui/addonStoreGui/controls/messageDialogs.py <https://github.com/nvaccess/nvda/pull/16636/files#diff-04af6494898544d5d6a576803476bd6fa843ff3683b467c7f83cc0d78facb61b>  (62) 
*	M source/gui/addonStoreGui/controls/storeDialog.py <https://github.com/nvaccess/nvda/pull/16636/files#diff-b21143677b5bc73364bf3f6dd6a9713720133959c6793aaf265e509fd1b8e91b>  (17) 
*	A source/utils/schedule.py <https://github.com/nvaccess/nvda/pull/16636/files#diff-e3259aaf59da07b3b21e20c04f11b41e7bb7c0eaf9b2855d8f4c32910f4213b0>  (61) 


Patch Links:


*	https://github.com/nvaccess/nvda/pull/16636.patch
*	https://github.com/nvaccess/nvda/pull/16636.diff

—
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/16636> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEHZHEKE2THTDDDVIY3ZE7753AVCNFSM6AAAAABISDQYZKVHI2DSMVQWIX3LMV43ASLTON2WKOZSGMZDMOJWG42DQOI> .
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

seanbudd commented:
@josephsl - not yet please, this is still an uncomplete draft

hwf1324 commented on requirements.txt:
```diff
@@ -17,7 +18,7 @@ fast-diff-match-patch==2.1.0
 typing-extensions==4.9.0 
```
```suggestion
typing-extensions==4.9.0
```

Also delete the spaces here.

seanbudd committed:
Update requirements.txt

Co-authored-by: WangFeng Huang <1398969445@qq.com>

Adriani90 commented:
Really cool to see NV Access starting work on this. Thanks Sean for this great contribution.

seanbudd committed:
minor fixup

seanbudd committed:
Add settings and docs

seanbudd committed:
Merge branch 'updateAddons' of https://github.com/nvaccess/nvda into updateAddons

seanbudd committed:
Fix up name for AddonsAutomaticUpdate

seanbudd committed:
make scheduling safer

seanbudd committed:
fixup typing

seanbudd committed:
add unit tests for scheduler

seanbudd committed:
add add-on list to dialog

seanbudd committed:
implement updating add-ons

seanbudd commented:
@coderabbitai review

seanbudd commented:
This is now ready for review and testing

seanbudd committed:
fix up headings

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -19,6 +19,10 @@ Unicode CLDR has been updated.
+* By default, after NVDA startup, the Add-on Store will notify you if any add-on updates are available. (#15035)
+  * This can be disabled in the "Add-on Store" section of settings.
+  * The notification checks daily for add-on updates.
+  * Only updates within the same channel will be checked (e.g. installed beta add-ons will only notify for updates to beta add-ons).
```
This is client side not server side, is this still an issue?

seanbudd replied:
addonStore.network.BASE_URL ensures communication is over https. The task scheduling doesn't interact with networking directly

seanbudd replied:
SHAs are used to verify downloads of specific add-ons

seanbudd commented on tests/manual/addonStore.md:
```diff
@@ -161,10 +161,62 @@ For example: "Clock".
+1. Using the context menu, install the add-ons.
+1. Exit the dialog
+1. Restart NVDA as prompted.
+1. Confirm the up-to-date add-ons are listed in the installed add-ons tab of the Add-ons Store.
```
The first 3 steps clarify the environment, could you highlight which part is unclear?

seanbudd replied:
This is clarified in step 4

seanbudd committed:
simplify message dialog

seanbudd commented on source/gui/addonStoreGui/viewModels/store.py:
```diff
@@ -58,21 +58,24 @@
 
 
 class AddonStoreVM:
+	onDisplayableError = DisplayableError.OnDisplayableErrorT()
+	"""
+	An extension point used to notify the add-on store VM when an error
+	occurs that can be displayed to the user.
+
+	This allows the add-on store GUI to handle displaying an error.
+
+	@param displayableError: Error that can be displayed to the user.
+	@type displayableError: gui.message.DisplayableError
+	"""
+
+	_downloader = AddonFileDownloader()
```
I'm considering this as out of scope

seanbudd commented on tests/manual/addonStore.md:
```diff
@@ -161,10 +161,62 @@ For example: "Clock".
+1. Test various buttons:
+    - Press "Update All": Ensure NVDA installs the add-ons.
+    - Press "Close": Ensure if Add-ons have been installed that NVDA prompts for restart afterwards
+    - Press "Open Add-on Store": Ensure NVDA opens to the Updatable tab in the Add-on Store
```
These are already included

seanbudd committed:
add to release blurb

seanbudd committed:
disable in secure mode

Qchristensen reviewed (approved):
User docs changes read well.

CyrilleB79 commented on source/config/configFlags.py:
```diff
@@ -202,3 +202,20 @@ def _displayStringLabels(self):
+			# Translators: This is a label for the automatic update behaviour for add-ons.
+			# It will notify the user when updates are available.
+			self.NOTIFY: _("Notify (default)"),
+			# self.UPDATE: _("Update Automatically"),
```
Translators comment please.

By the way, why hasn't it been captured by the check script?

seanbudd replied:
This is commented out

CyrilleB79 reviewed (commented):
First review (only partial)

nvdaes commented:
I've test this and, when an external add-on is  installed,  I get the following error, not present wwithout external add-ons:

> NVDA initialized
ERROR - unhandled exception (18:37:01.668) - MainThread (8896):
Traceback (most recent call last):
  File "C:\Users\nrm19\Documents\repos\nvda\.venv\Lib\site-packages\wx\core.py", line 3427, in <lambda>
    lambda event: event.callable(*event.args, **event.kw) )
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "gui\addonStoreGui\controls\messageDialogs.py", line 545, in _checkForUpdatableAddons
    addonsPendingUpdate = addonDataManager._addonsPendingUpdate()
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "addonStore\dataManager.py", line 319, in _addonsPendingUpdate
    if addon.channel == addon._addonHandlerModel._addonStoreData.channel:
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'channel'


Also, the dialog notifying about updates is shown in a Window not very prominent, and no alert is presented, so I realised that it was there casually, after pressing alt+tab.

Thanks for this amazing feature!

nvdaes commented on user_docs/en/changes.md:
```diff
@@ -59,6 +64,9 @@ It is especially useful to read the error location markers in tracebacks. (#1632
 * When a `gainFocus` event is queued with an object that has a valid `focusRedirect` property, the object pointed to by the `focusRedirect` property is now held by `eventHandler.lastQueuedFocusObject`, rather than the originally queued object. (#15843)
 * NVDA will log its executable architecture (x86) at startup. (#16432, @josephsl)
 * `wx.CallAfter`, which is wrapped in `monkeyPatches/wxMonkeyPatches.py`, now includes proper `functools.wraps` indication. (#16520, @XLTechie)
+* There is a new module for scheduling tasks, using the pip module `schedule`.
```
I think that some advice to schedule jobs that display messages undefinitely at intervals would be useful. I'm thinking in add-ons like Gestor de Pomodoros, by @jpavonabian . Imo the schedule module maybe helpful, but some suggestions to avoid conflict, or a small example in the developer guide, would be appreciated. For example, what happens if a message is presented each 30 seconds and another job overlaps this?

seanbudd replied:
I've improved the function descriptions and API a bit further. Can you let me know your thoughts of what else could change?

nvdaes commented:
Also, shouldn't be the addonStore configuration section in BASE_ONLY_SECTIONS of config.conf?
If this will apply just when starting NVDA and it's changed while a configuration profile is not active, this maybe missed.

nvdaes commented on tests/manual/addonStore.md:
```diff
@@ -161,10 +161,63 @@ For example: "Clock".
+    - For installed copies: `%APPDATA%\nvda`
+1. To spoof an old release, we need to edit 2 files:
+    - Add-on store JSON metadata
+        - Located in: `.\addonStore\addons\`.
```
```suggestion
        - Located in: `\.addons\`.

```

nvdaes commented on tests/manual/addonStore.md:
```diff
@@ -161,10 +161,63 @@ For example: "Clock".
+1. To spoof an old release, we need to edit 2 files:
+    - Add-on store JSON metadata
+        - Located in: `.\addonStore\addons\`.
+        - Example: `source\userConfig\addonStore\addons\clock.json`
```
```suggestion
        - Example: `source\userConfig\addons\clock.json`

```

nvdaes commented on tests/manual/addonStore.md:
```diff
@@ -161,10 +161,63 @@ For example: "Clock".
+    - For installed copies: `%APPDATA%\nvda`
+1. To spoof an old release, we need to edit 2 files:
+    - Add-on store JSON metadata
+        - Located in: `.\addonStore\addons\`.
```
```suggestion
        - Located in: `.\addons\`.

```

nvdaes commented on tests/manual/addonStore.md:
```diff
@@ -161,10 +161,63 @@ For example: "Clock".
+1. To spoof an old release, we need to edit 2 files:
+    - Add-on store JSON metadata
+        - Located in: `.\addonStore\addons\`.
+        - Example: `source\userConfig\addonStore\addons\clock.json`
```
```suggestion
        - Example: `source\userConfig\addons\clock.json`

```

nvdaes commented:
The close button of the Add-on updates available dialog is not associated with Escape. Is this intentional?

nvdaes commented:
When the Update all add-ons is pressed, the Close button is focused. Shouldn't be more consistent focusing the add-ons list?

seanbudd committed:
improve the scheduleThread module and functionality

seanbudd committed:
review feedback

seanbudd committed:
minor fixups

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -5,7 +5,8 @@
 This release adds support for Unicode Normalization to speech and braille output.
 This can be useful when reading characters that are unknown to a particular speech synthesizer or braille table and which have a compatible alternative, like the bold and italic characters commonly uses on social media.
 It also allows reading of equations in the Microsoft Word equation editor.
-You can enable this functionality for both speech and braille in their respective settings categories in the NVDA Settings dialog.
```
I think it should be removed now that more content is in the release blurb

seanbudd committed:
set focus and raise

seanbudd commented:
@nvdaes  - thanks for your testing

> I've tested this and, when an external add-on is installed, I get the following error, not present wwithout external add-ons:
> 
> Also, the dialog notifying about updates is shown in a Window not very prominent, and no alert is presented, so I realised that it was there casually, after pressing alt+tab.

These issues should be fixed now.

>Also, shouldn't be the addonStore configuration section in BASE_ONLY_SECTIONS of config.conf?
If this will apply just when starting NVDA and it's changed while a configuration profile is not active, this maybe missed.

Good catch, fixed.

>
The close button of the Add-on updates available dialog is not associated with Escape. Is this intentional?

I think this is typical for this kind of message dialogs? No special behaviour has been written here.

>
When the Update all add-ons is pressed, the Close button is focused. Shouldn't be more consistent focusing the add-ons list?

Fixed this to focus the first list item

seanbudd reviewed (commented):
@CyrilleB79 thanks for your review

seanbudd commented:
@coderabbitai review

seanbudd commented:
https://github.com/coderabbitai review

seanbudd commented:
@coderabbitai review

seanbudd commented on tests/unit/test_util/test_schedule.py:
```diff
@@ -0,0 +1,137 @@
+			# Call the scheduleJob method with the same cron time
+			ScheduleThread.scheduleJob(jobFunc, jobSchedule, ThreadTarget.GUI)
+
+	def test_calculateDailyTimeOffset(self):
+		todayAtMidnight = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
+		NVDAState.getStartTime = MagicMock(return_value=todayAtMidnight.timestamp())
+		# Call the _calculateDailyTimeOffset method
+		offset = ScheduleThread._calculateDailyTimeOffset()
+		# Assert that the offset is calculated correctly
+		self.assertEqual(offset, f"00:{ScheduleThread.DAILY_JOB_MINUTE_OFFSET:02d}")
+		
+		ScheduleThread.scheduledDailyJobCount = 1
+		offset = ScheduleThread._calculateDailyTimeOffset()
+		self.assertEqual(offset, f"00:{ScheduleThread.DAILY_JOB_MINUTE_OFFSET * 2:02d}")
+
+		# Test the case where the start time is 11:59 to ensure the hour offset is calculated correctly
+		NVDAState.getStartTime = MagicMock(return_value=todayAtMidnight.replace(hour=11, minute=59).timestamp())
+		ScheduleThread.scheduledDailyJobCount = 0
+		offset = ScheduleThread._calculateDailyTimeOffset()
+		expectedMinOffset = (ScheduleThread.DAILY_JOB_MINUTE_OFFSET + 59) % 60
+		self.assertEqual(offset, f"12:{expectedMinOffset:02d}")
+
+		# Test the case where the start time is 23:59 to ensure the day and hour offset is calculated correctly
+		NVDAState.getStartTime = MagicMock(return_value=todayAtMidnight.replace(hour=23, minute=59).timestamp())
+		ScheduleThread.scheduledDailyJobCount = 0
+		offset = ScheduleThread._calculateDailyTimeOffset()
+		expectedMinOffset = (ScheduleThread.DAILY_JOB_MINUTE_OFFSET + 59) % 60
+		self.assertEqual(offset, f"00:{expectedMinOffset:02d}")
```
this is not relevant as we are scheduling daily jobs and adding a fixed referenced time, so the times are predictable.

seanbudd commented on tests/unit/test_util/test_schedule.py:
```diff
@@ -0,0 +1,137 @@
+		scheduledJob = ScheduleThread.scheduledJobs[0]
+		self.assertEqual(f"{scheduledJob.at_time.hour:02d}:{scheduledJob.at_time.minute:02d}", cronTime)
+
+	def test_scheduleJob_jobClash(self):
+		def jobFunc():
+			# Job function implementation
+			pass
+
+		cronTime = "12:00"  # Example cron time
+		jobSchedule = schedule.every().day.at(cronTime)
+
+		# Call the scheduleJob method
+		ScheduleThread.scheduleJob(jobFunc, jobSchedule, ThreadTarget.GUI)
+
+		with self.assertRaises(JobClashError):
+			# Call the scheduleJob method with the same cron time
+			ScheduleThread.scheduleJob(jobFunc, jobSchedule, ThreadTarget.GUI)
+
```
There is clean up logic in `tearDown`

seanbudd replied:
yes thats correct

seanbudd commented on tests/unit/test_util/test_schedule.py:
```diff
@@ -0,0 +1,137 @@
+		def incrementC(scheduledVals: list):
+			scheduledVals[2] += 1
+
+		ScheduleThread.scheduleDailyJobAtStartUp(incrementA, ThreadTarget.CUSTOM, scheduledVals)
+		ScheduleThread.scheduleDailyJobAtStartUp(incrementB, ThreadTarget.CUSTOM, scheduledVals)
+		ScheduleThread.scheduleDailyJobAtStartUp(incrementC, ThreadTarget.CUSTOM, scheduledVals)
+		# Sanity checks (have failed in development)
+		self.assertEqual(len(ScheduleThread.scheduledJobs), 3)
+		self.assertEqual(ScheduleThread.scheduledDailyJobCount, 3)
+
+		expectedResult = [0, 0, 0]
+		self.assertEqual(scheduledVals, expectedResult)
+		for jobIndex in range(3):
+			startTime = NVDAState.getStartTime()
+			currentJob = ScheduleThread.scheduledJobs[jobIndex]
+
+			# Ensure that the job is scheduled to run at the expected time
+			expectedSecsOffsetMin = jobIndex * ScheduleThread.DAILY_JOB_MINUTE_OFFSET * 60
+			expectedSecsOffsetMax = (jobIndex + 1) * ScheduleThread.DAILY_JOB_MINUTE_OFFSET * 60
+			nextRun = currentJob.next_run.timestamp()
+			actualSecsOffset = nextRun - startTime
+			self.assertLessEqual(
+				actualSecsOffset,
+				expectedSecsOffsetMax,
+				f"Job {jobIndex} was not scheduled as expected. Job: {currentJob}"
+			)
+			self.assertGreaterEqual(
+				actualSecsOffset,
+				expectedSecsOffsetMin,
+				f"Job {jobIndex} was not scheduled as expected. Job: {currentJob}"
+			)
+
+			# Ensure the job runs as expected
+			currentJob.run()
+			expectedResult[jobIndex] += 1
+			self.assertEqual(
+				scheduledVals,
+				expectedResult,
+				f"Job {jobIndex} did not run as expected. Scheduled jobs: {ScheduleThread.scheduledJobs}"
+			)
```
Could you provide example assertions?

seanbudd commented on tests/unit/test_util/test_schedule.py:
```diff
@@ -0,0 +1,137 @@
+				f"Job {jobIndex} did not run as expected. Scheduled jobs: {ScheduleThread.scheduledJobs}"
+			)
+
+	def test_scheduleJob(self):
+		def jobFunc():
+			# Job function implementation
+			pass
+
+		cronTime = "12:00"  # Example cron time
+		jobSchedule = schedule.every().day.at(cronTime)
+
+		# Call the scheduleJob method
+		ScheduleThread.scheduleJob(jobFunc, jobSchedule, ThreadTarget.GUI)
+
+		# Assert that the job is scheduled correctly
+		self.assertEqual(len(ScheduleThread.scheduledJobs), 1)
+		scheduledJob = ScheduleThread.scheduledJobs[0]
+		self.assertEqual(f"{scheduledJob.at_time.hour:02d}:{scheduledJob.at_time.minute:02d}", cronTime)
+
```
this is done here, cronTime is fixed

seanbudd commented on source/addonStore/dataManager.py:
```diff
@@ -308,6 +309,21 @@ def _getCachedInstalledAddonData(self, addonId: str) -> Optional[InstalledAddonS
 			return None
 		return _createInstalledStoreModelFromData(cacheData)
 
+	def _addonsPendingUpdate(self) -> list["_AddonGUIModel"]:
+		addonsPendingUpdate: list["_AddonGUIModel"] = []
+		compatibleAddons = self.getLatestCompatibleAddons()
+		for channel in compatibleAddons:
+			for addon in compatibleAddons[channel].values():
+				if (
+					getStatus(addon, _StatusFilterKey.UPDATE) == AvailableAddonStatus.UPDATE
+					# Only consider add-ons that have been installed through the Add-on Store
+					and addon._addonHandlerModel._addonStoreData is not None
+				):
+					# Only consider add-on updates for the same channel
+					if addon.channel == addon._addonHandlerModel._addonStoreData.channel:
+						addonsPendingUpdate.append(addon)
+		return addonsPendingUpdate
```
this refactor is considerably more complex than the current code

seanbudd committed:
fix lint

wmhn1872265132 commented:
Can the Add-ons of the External channel be notified of updates?

wmhn1872265132 commented:
Is it possible to add a menu item or command that manually triggers checking for Add-ons updates to facilitate testing?
Not sure if you need this feature for daily use.

nvdaes commented on source/gui/addonStoreGui/__init__.py:
```diff
@@ -1,10 +1,21 @@
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
 
+from utils.schedule import ScheduleThread, ThreadTarget
```
```suggestion
from utils.schedule import scheduleThread, ThreadTarget

```

nvdaes commented on source/gui/addonStoreGui/__init__.py:
```diff
@@ -1,10 +1,21 @@
+
+
+def initialize():
+	ScheduleThread.scheduleDailyJobAtStartUp(
```
```suggestion
	scheduleThread.scheduleDailyJobAtStartUp(

```

nvdaes commented:
Shouldn't be the scheduleThread global variable used instead of the ScheduleTread class, to handle job clashes?
Sorry if I'm missing something here.

nvdaes commented on user_docs/en/changes.md:
```diff
@@ -59,6 +63,10 @@ It is especially useful to read the error location markers in tracebacks. (#1632
 * NVDA will log its executable architecture (x86) at startup. (#16432, @josephsl)
 * `wx.CallAfter`, which is wrapped in `monkeyPatches/wxMonkeyPatches.py`, now includes proper `functools.wraps` indication. (#16520, @XLTechie)
+* There is a new module for scheduling tasks `utils.schedule`, using the pip module `schedule`. (#16636)
+  * You can use `ScheduleThread.scheduleDailyJobAtStartUp` to automatically schedule a job that happens after NVDA starts, and every 24 hours after that.
```
```suggestion
  * You can use `scheduleThread.scheduleDailyJobAtStartUp` to automatically schedule a job that happens after NVDA starts, and every 24 hours after that.

```

nvdaes commented on user_docs/en/changes.md:
```diff
@@ -59,6 +63,10 @@ It is especially useful to read the error location markers in tracebacks. (#1632
+* There is a new module for scheduling tasks `utils.schedule`, using the pip module `schedule`. (#16636)
+  * You can use `ScheduleThread.scheduleDailyJobAtStartUp` to automatically schedule a job that happens after NVDA starts, and every 24 hours after that.
+  Jobs are scheduled with a delay to avoid conflicts.
+  * `ScheduleThread.scheduleDailyJob` and `scheduleJob` can be used to schedule jobs at custom times, where a `JobClashError` will be raised on a known job scheduling clash.
```
```suggestion
  * `scheduleThread.scheduleDailyJob` and `scheduleJob` can be used to schedule jobs at custom times, where a `JobClashError` will be raised on a known job scheduling clash.

```

nvdaes commented:
I think that, if the usage of ScheduleThread class is encouraged, instead of the scheduleThread global variable, NVDA won't be able to handle conflicting jobs. Right?

nvdaes commented:
> When the Update all add-ons is pressed, the Close button is focused. Shouldn't be more consistent focusing the add-ons list?

@seanbudd, this seems not to be fixed, though I think it's a minor issue.

Also, I think that you may want to add a classmethod in the ScheduleThread class to cancel specific jobs, removing them from the scheduledJobs list. This method may include a tag, which authors may want to associate with add-on ids, and jobs maybe removed one by one or filtering them by tags. I think that add-ons may use this and  this may avoid conflicts.
See the Cancel several jobs, filtered by tags, in the [Schedule examples](https://schedule.readthedocs.io/en/stable/examples.html).

I've created two jobs to say Hello and Bye in the console, using ui.message, and since they were overlapped sometimes they were missed, but the first job was run as expected, with some exceptions related to UIAHandler when opening files, while messages should be presented also in braille. But generally this has worked as expected.

nvdaes commented on source/utils/schedule.py:
```diff
@@ -0,0 +1,189 @@
+	"""
+	No thread target.
+	Runs directly and blocks `scheduleThread`.
+	Run code on a custom thread to ensure `scheduleThread` is not blocked.
```
Here, for me it's not clear if custom thread means a thread created using a class derived from the ScheduleThread class, without using the scheduleThread global variable, or if this means to use the ThreadTarget.CUSTOM with the scheduleThread global variable.

seanbudd replied:
I've added better docs now, does this help?

seanbudd committed:
remove double handling of tracking jobs

seanbudd committed:
handle review comments

seanbudd committed:
review docs

seanbudd commented:
@wmhn1872265132

>Can the Add-ons of the External channel be notified of updates?

No, this only supports add-ons installed from the add-on store, within the same channel.

>Is it possible to add a menu item or command that manually triggers checking for Add-ons updates to facilitate testing?
Not sure if you need this feature for daily use.

You should be able to do it from the python console using the manual test plan steps in this PR. does this solve the problem you have with testing?

@nvdaes - thanks for your review

>Shouldn't be the scheduleThread global variable used instead of the ScheduleTread class, to handle job clashes?
Sorry if I'm missing something here.

Good catch, yes, it is safer. I've adjusted code to use classmethods so that you now must use the variable not the class

>I think that, if the usage of ScheduleThread class is encouraged, instead of the scheduleThread global variable, NVDA won't be able to handle conflicting jobs. Right?

Not quite, their behaviour should be identical. The issue is if you use the class you could potentially have no running thread, i.e. between initialisation and termination, which could cause bad behaviour.

>Also, I think that you may want to add a classmethod in the ScheduleThread class to cancel specific jobs, removing them from the scheduledJobs list. 

I've removed the scheduledJobs list. I'd encourage authors to directly use `schedule.jobs` to manage jobs.

>When the Update all add-ons is pressed, the Close button is focused. Shouldn't be more consistent focusing the add-ons list?
>this seems not to be fixed, though I think it's a minor issue.

I think is fixed again apologies

seanbudd commented:
@coderabbitai

seanbudd committed:
fix lint

wmhn1872265132 commented:
1. The column title of the update dialog box has an extra `:`
2. The Name in the update dialog box displays the ID of the Add-on, and the Add-on Store displays the summary of the Add-on.

Here is a sample output:

Update dialog:
> TransIPA; Installed version:: 1.2; Available version:: 1.2.2; Channel:: Stable; Status:: Update Available

Add-on Store
> Reporting the International Phonetic Alphabet in Braille; Status: Update Available; Installed version: 1.2; Available version: 1.2.2; Channel: Stable; Publisher: Cary-rowen

seanbudd committed:
address review comments

seanbudd commented:
@wmhn1872265132 - those changes have been fixed up

wmhn1872265132 commented:
> The Name in the update dialog box displays the ID of the Add-on, and the Add-on Store displays the summary of the Add-on.

This issue is still unresolved

seanbudd committed:
address review comments

seanbudd commented:
Thanks - I missed that comment

XLTechie commented on tests/manual/addonStore.md:
```diff
@@ -100,38 +100,40 @@ You can do this by:
-### Updating from add-on installed from add-on store
-1. [Install an add-on from the add-on store](#install-add-on)
+### Simulate creating an updatable add-on from the Add-on Store
+Without having a previously installed add-on get an update, it is hard to naturally test updatable add-ons.
```
This wording is awkward.

```suggestion
Without having an installed add-on which has an update pending, it is hard to test updatable add-ons.
```

XLTechie commented on tests/manual/addonStore.md:
```diff
@@ -100,38 +100,40 @@ You can do this by:
-1. Open the add-on store
-1. Ensure the same add-on you edited is available on the add-on store with the status "update".
+
+### Updating from add-on installed from Add-on Store
```
In addition to the wording clarity suggestion below, I would like to echo the AI's opinion that we should have blank lines around headings. It makes reading easier, and is far more standard in non-NVA MD documents.

I would think it also made reading the source visually easier, but in reading it with speech, reading rapidly and hitting blank lines around headings helps them stand out in a large wall of speech.

```suggestion
### Updating from add-on originally installed via Add-on Store
```

nvdaes commented:
For me, now all is working as expected, and documentation for schedule is clear.
Note: in the updatableAddons dialog, when the list is focused, the first element is not selected.

Adriani90 commented:
Using portable copies of NVDA does not seem to work.
Testing as follows:
1. Downloaded the snapshot artifact
2. Created a portable copy of the snapshot and saved in a previously created folder under downloads\NVDA
3. Dowloanded and installed an older version (v2.6) of the add-on character information from here:
https://github.com/CyrilleB79/charInfo/releases/tag/V2.6
4. After NVDA restart, Checked in add-on store under updatable add-ons tab whether there is an update available for the add-on, which is the case (last version: 3.0)
5. Made sure the notification setting is set to the notify value

Actual behavior: Update notification does not appear after restarting the NVDA portable copy

Expected behavior: the notification should appear after restarting the NVDA portable copy.

Adriani90 commented:
Does this PR notify for add-on updates only when they are compatible?

Adriani90 commented:
Also it seems the notification is expected to work only for add-ons which are installed through the add-on store.
However, if I download the add-on mannually from its source (e.g. Github), after installation the update is still shown in the add-on store but the notification doesn't work. This will definitely cause confusions.
I suggest to allow notifications to be shown also when installing add-ons manually also from the source page of the add-on which is registered to the add-on store.

CyrilleB79 commented:
@Adriani90 please not that I have not released a version compatible with NVDA 2024.1 of this add-on. So in this case, the question is:
Do we want to be notified of newer incompatible release (e.g. 3.0) of currently incompatible add-ons (2.6)?

Adriani90 commented:
> So in this case, the question is: Do we want to be notified of newer incompatible release (e.g. 3.0) of currently incompatible add-ons (2.6)?

Definitely yes in my opinion. It is always better to have an improved incompatible newer version of an add-on rather than sticking to an old incompatible version.
If the author released a new incompatible version, then he or she still documented its changes, users would expect to be at least notified that there is a newer version, whether they install it or not.

CyrilleB79 commented:
> Also it seems the notification is expected to work only for add-ons which are installed through the add-on store. However, if I download the add-on mannually from its source (e.g. Github), after installation the update is still shown in the add-on store but the notification doesn't work. This will definitely cause confusions. I suggest to allow notifications to be shown also when installing add-ons manually also from the source page of the add-on which is registered to the add-on store.

The reason is probably that when you have an externally installed add-on, NVDA cannot know if it is a dev, beta or stable one. Actually, dev, beta or stable is only a property of the channel, not of the add-on itself.
Though, I agree that it will cause confusion.
At least, if not already, this limitation should be clearly documented.

CyrilleB79 commented:
> > So in this case, the question is: Do we want to be notified of newer incompatible release (e.g. 3.0) of currently incompatible add-ons (2.6)?
> 
> Definitely yes in my opinion. It is always better to have an improved incompatible newer version of an add-on rather than sticking to an old incompatible version. If the author released a new incompatible version, then he or she still documented its changes, users would expect to be at least notified that there is a newer version, whether they install it or not.

Yes, I agree.

Adriani90 commented:
> The reason is probably that when you have an externally installed add-on, NVDA cannot know if it is a dev, beta or stable one. Actually, dev, beta or stable is only a property of the channel, not of the add-on itself.

But when you install it NVDA could check its manifest file which should actually contain the proper version info. Or am I wrong?

Adriani90 commented:
Also @seanbudd could you please convert the notification setting to a checkbox?
Actually the convention in NVDA GUI is as follows:
* If a setting has only values "enabled" or "disabled", then it should be a checkbox unless it is a feature flag which belongs in advanced settings and the default value is a third value for testing purposes until the setting becomes stable. After it becomes stable, it is converted back to a checkbox and the third value for testing purposes is removed.
* In case there is no feature flag, A combo box is used for values with different labels other than "enabled" and "disabled" (e.g. braille modes combo box.

In this case, the label of the setting is already "notifications" so having a value with the same name is not needed.

Unfortunately e.g. in the braille settings category this convention has been broken by who ever implemented and reviewed those pull requests. And it seems we see this inconsistencies more often.
This made the GUI inconsistent and confusing for users who use these conventions (e.g. via key strokes. It would be nice if revierwers pay some attention on this in the future, otherwise the GUI will loose consistency more and more.

XLTechie commented:
I also agree with all of the various points raised by @Adriani90. The no updates for external but store-available add-ons, incompatible updates, and checkboxes in the GUI.

These all have the probability to be confusing to users and generate a lot of back and forth (maybe a bit less so the checkbox thing, but I am annoyed that feature flags are showing up in Speech and other panels).

Commenting on one specific thing that  @Adriani90 wrote:

> But when you install it NVDA could check its manifest file which should actually contain the proper version info.

I think this might need a separate feature request, but I agree the Store should "adopt" external add-ons that match known store add-ons. There may be security concerns with doing that, however.

seanbudd commented:
@Adriani90 

> Using portable copies of NVDA does not seem to work.

This appears to be relying on a add-on installed externally. This is not supported in this iteration. Can you please confirm with an add-on installed through the add-on store?

>Does this PR notify for add-on updates only when they are compatible? It seems so when looking at the code.
I think it is really important that NVDA notifies users also of updates to add-ons which are currently incompatible.
>If the author released a new incompatible version, then he or she still documented its changes, users would expect to be at least notified that there is a newer version, whether they install it or not.

Add-ons which are currently incompatible but have a compatible available update should get notifications.
Add-ons with an incompatible update need to be updated through the store, however this is something we could reconsider in future iterations.

>I suggest to allow notifications to be shown also when installing add-ons manually also from the source page of the add-on which is registered to the add-on store.

This can't be supported in the current implementation as we are only providing update notifications for add-ons within the same channel. Allowing cross-channel updates is out of scope of this iteration.

>could you please convert the notification setting to a checkbox?

We are using a combox box in this situation as a third state will be added in future, please see the PR description.

seanbudd committed:
fix lint and focus

seanbudd commented on tests/manual/addonStore.md:
```diff
@@ -100,38 +100,40 @@ You can do this by:
 1. Go to your NVDA user configuration folder:
-    - For source: `.\source\userConfig`
-    - For installed copies: `%APPDATA%\nvda`
+    - For source: `.\source\userConfig\addons`
```
I'll add reference to portable copies

seanbudd committed:
Apply suggestions from code review

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

seanbudd committed:
improve test plan docs

seanbudd commented:
>But when you install it NVDA could check its manifest file which should actually contain the proper version info. Or am I wrong?

This is the case generally, which is why you can update external add-ons using the add-on store in many cases, with the following exceptions:
- the add-on manifest doesn't support channel information yet, which means we cannot guarantee the current channel for same channel. This isn't an issue if you don't care about the current channel of the add-on. If we introduce more flexible channel update options, auto-updates will handle external add-ons fine.
- the add-on manifest has a version which we cannot parse. in many cases, strange version number strings exist in the manifest. we only allow major.minor.patch versioning in the store, which means different or invalid versioning system will make it impossible to know if the version available is an upgrade or downgrade. in these situations we have a "replace" button. 

One way around this in future could be @XLTechie's idea
>I think this might need a separate feature request, but I agree the Store should "adopt" external add-ons that match known store add-ons. There may be security concerns with doing that, however.

This could be possible if we make an add-on store endpoint to fetch the add-on store version data given the SHA of an add-on.
This would be potentially a large project with how the add-on store is set up.

nvdaes commented:
I`'ve reviewed changes and tested dialogs. I think that this can be merged for the first iteration, so it can be widely tested, including the schedule module in add-ons under the dev channel.
Note that the first element of the list of updatable add-ons in the UpdatableAddonsDialog, which contains buttons to open the store and Update all, is not focused, but I think that this is hard to achieve, even using `Focus(0)`. Imo this shouldn't be a problem since add-ons can be reviewed with down arrow, and this shouldn't block this PR to be merged.

CyrilleB79 commented:
The feature is nice and I think that this can ship as is to be tested in alpha to get first public feedback.

Though, I can see these little points that may be fixed/updated:
1. In the "Add-on updates available", there is a "State" column. But the state seems to be always "Update available". Or is there another possibility? If not, I'd remove this unuseful column.
2. The Add-on Store settings panel indicates the profile name but I think that the notification parameter is only saved in the base config. So the title should not change (as per general settings)
3. I have only two items in the "Update notifications" combo-box: "Notify (default)" and "Disabled". I would also expect "Notify" as done with with feature flags. Either you use a feature flag and use 3 values (Default (Notify), Notify and Disabled), or you just use a standard combo-box without indicating what the default is.

seanbudd committed:
do not announce profile change

seanbudd committed:
remove default from label

seanbudd commented:
@CyrilleB79  thanks for looking into this:
1. The state updates as add-ons are downloaded and installed, like within the add-on store
1. I think I've fixed up profile announcements, please confirm
1. We are using a combox box in this situation as a third state will be added in future, please see the PR description. I can remove the "default" part of the label

Qchristensen reviewed (approved):
Looks good from docs point of view.  Great work!

seanbudd commented:
@coderabbitai review

CyrilleB79 commented on source/gui/addonStoreGui/controls/messageDialogs.py:
```diff
@@ -348,3 +357,203 @@ def __init__(self, parent: wx.Window):
+		evt.Skip()
+
+	def _setupMessage(self, sHelper: BoxSizerHelper):
+		_message = pgettext(
+			"addonStore",
+			# Translators: Message displayed when updates are available for some installed add-ons.
+			"Updates are available for some of your installed add-ons. "
+		)
```
@coderabbitai, the `pgettext` function does not need to be imported. It is made available in `languageHandler.setLanguage` by installing the gettext translation.
See https://docs.python.org/3/library/gettext.html for more details.

CyrilleB79 commented:
@seanbudd:

>     1. The state updates as add-ons are downloaded and installed, like within the add-on store

OK. Makes sense, especially when the download time is long.

>     2. I think I've fixed up profile announcements, please confirm

Confirmed. Thanks!

>     3. We are using a combox box in this situation as a third state will be added in future, please see the PR description. I can remove the "default" part of the label

Yes, it should be removed. But it is actually already.


@nvdaes, I agree to add in the settings panel (at the end) the checkbox corresponding to the "Don't show me again" checkbox of the add-on store. Though, it is not mandatory to discuss and add it in the same PR.

nvdaes commented:
Cyrille wrote:

> @nvdaes, I agree to add in the settings panel (at the end) the checkbox corresponding to the "Don't show me again" checkbox of the add-on store. Though, it is not mandatory to discuss and add it in the same PR.

I agree: it can be added later and discussed separately. I think that this PR should be tested with enough time, not just for its main feature, but also for the schedule module. I'll release a dev version ofthe readFeedsadd-on using it. Personally, I find anoying the dialogs presented by add-ons notifying about updates, and I think that this PR will be very appreciated.

CyrilleB79 commented:
Another improvement could be to give the choice between a notification dialog (as used currently) and a toast notification which is less invasive. Add-on Updater already give this choice and toast notification is appreciated, even if it is not suitable for everyone.

But again, this improvement can be discussed and implemented in a subsequent PR.

Adriani90 commented:
Thanks Sean, yes I agree my points can be discussed in the next iteration. One point though:
You wrote:
> As mentioned in the other discussions for this, that checkbox only hides incompatible update versions.

Even incompatible update versions should not be hidden by default. It is always better to know there is a newer version of an add-on even if the newer version is still incompatible. The add-on author documented the changes for that version, so in my view all available updates should show up in the updatable add-ons tab regardless whether they are compatible or not. The users can then decide which updates they want to install.

nvdaes commented:
> Another improvement could be to give the choice between a notification dialog (as used currently) and a toast notification which is less invasive. Add-on Updater already give this choice and toast notification is appreciated, even if it is not suitable for everyone.
> 
> But again, this improvement can be discussed and implemented in a subsequent PR.

Yes, I agree, though also ui.message maybe used in addition to notifications. Another improvement would be to show new (not installed) add-ons published on the store, like my bot used on Mastodon, my spanish mailing list and the feeds subgroup. I plan to create an issue to request this, and if it's not accepted, an add-on maybe created with a button to open the store to install new add-ons and see its description.
I know that these comments maybe hidden, but they maybe included using the same add-on dialogs of this PR, so maybe considered in the design of current dialogs, for example, to support the decission of keeping the updatable state shown in the updatable add-ons dialog, since the available state, if new add-ons can be presented in the future, can be included to distinguish between new and installed/updatable add-ons.

seanbudd commented on tests/manual/addonStore.md:
```diff
@@ -100,38 +102,39 @@ You can do this by:
 For example: "Clock".
-1. Go to your NVDA user configuration folder:
-    - For source: `.\source\userConfig`
-    - For installed copies: `%APPDATA%\nvda`
-1. To spoof an old release, we need to edit 2 files:
-    - Add-on store JSON metadata
-        - Located in: `.\addonStore\addons\`.
-        - Example: `source\userConfig\addonStore\addons\clock.json`
+1. Go to the "addons" folder in your [NVDA user configuration folder](#editing-user-configuration)
+1. To mock an old release, we need to edit 2 files:
+    - Add-on Store JSON metadata
+        - Example: `source\userConfig\addons\clock.json`
```
yes please

seanbudd commented on source/config/configFlags.py:
```diff
@@ -202,3 +202,20 @@ def _displayStringLabels(self):
+
+class AddonsAutomaticUpdate(DisplayStringStrEnum):
+	NOTIFY = "notify"
+	# UPDATE = "update"
```
```suggestion
	# TODO: uncomment when implementing #3208
	# UPDATE = "update"
```

seanbudd committed:
Update source/config/configFlags.py

seanbudd commented:
Can we please keep conversation on topic for this PR, and create new issues for these. It's adding a lot of noise to the discussion here.

seanbudd committed:
add script to update add-on files

nvdaes reviewed (approved):
I've seen the code for commits made today and also I've tested the Powershell scripts. For me it works as expected.
